src/resmom/mom_comm.c

```diff
@@ -3941,6 +3954,250 @@ im_request(int stream, int version)
+				if (ret != DIS_SUCCESS) {
+					sprintf(log_buffer, "SPAWN_MULTI read of argc");
+					free(node_list);
+					goto err;
+				}
+				argv = (char **)calloc(argc+1, sizeof(char *));
+				assert(argv);
+				for (i=0; i<argc; i++) {
+					argv[i] = disrst(stream, &ret);
+					if (ret != DIS_SUCCESS)
+						break;
+				}
+			}
+			argv[i] = NULL;
+			if (ret != DIS_SUCCESS) {
+				arrayfree(argv);
+				free(node_list);
+				sprintf(log_buffer, "SPAWN_MULTI read of argv array");
+				goto err;
+			}
+
+			num = 8;
+			envp = (char **)calloc(sizeof(char *), num);
+			assert(envp);
+			for (i=0;; i++) {
+				if ((cp = disrst(stream, &ret)) == NULL)
+					break;
+				if (ret != DIS_SUCCESS)
+					break;
+				if (*cp == '\0') {
+					free(cp);
+					break;
+				}
+				if (i == num-1) {
+					num *= 2;
+					envp = (char **)realloc(envp,
+						num*sizeof(char *));
+					assert(envp);
+				}
+				envp[i] = cp;
+			}
+			envp[i] = NULL;
+			if (ret != DIS_EOD) {
+				arrayfree(argv);
+				arrayfree(envp);
+				free(node_list);
+				sprintf(log_buffer, "SPAWN_MULTI read of envp array");
+				goto err;
+			}
+
+			/* Allocate a temporary place to store task IDs */
+                	taskidlist = (tm_task_id *)calloc(list_size, sizeof(tm_task_id));
+			assert(taskidlist);
+
+			/*
+			** Go through the list of nodes and do something
+			** when it's me, otherwise keep looking
+			*/
+			for (i = 0; i < list_size; i++) {
+				if (pjob->ji_nodeid != TO_PHYNODE(node_list[i])) {
+					continue;
+				}
+					
+#ifdef PMIX
+				pbs_pmix_register_client(pjob, node_list[i], &envp);
+#endif
+				if ((ptask = momtask_create(pjob)) == NULL) {
+					sprintf(log_buffer, "momtask_create "
+						"failed to create a new task\n");
+					log_event(PBSEVENT_DEBUG3,
+						PBS_EVENTCLASS_JOB, LOG_DEBUG,
+						jobid, log_buffer);
+					break;
+				}
+				strcpy(ptask->ti_qs.ti_parentjobid, jobid);
+				ptask->ti_qs.ti_parentnode = pvnodeid;
+				ptask->ti_qs.ti_myvnode    = node_list[i];
+				ptask->ti_qs.ti_parenttask = fromtask;
+				if (task_save(ptask) == -1) {
+					sprintf(log_buffer, "unable to save "
+						"task info\n");
+					log_event(PBSEVENT_DEBUG3,
+						PBS_EVENTCLASS_JOB, LOG_DEBUG,
+						jobid, log_buffer);
+					break;
+				}
+				errcode = start_process(ptask, argv, envp, false);
+				if (errcode != PBSE_NONE) {
+					sprintf(log_buffer, "start_process "
+						"failed with errcode %d\n",
+						errcode);
+					log_event(PBSEVENT_DEBUG3,
+						PBS_EVENTCLASS_JOB, LOG_DEBUG,
+						jobid, log_buffer);
+					break;
+				} else {
+					/* store off the taskid to send all at once */
+					taskidlist[created] = ptask->ti_qs.ti_task;
+					created++;
+				}
+			} /* end of for loop of node_list */
+
+			/* We finished going through the node_list, now send im_compose IM_ALL_OKAY */
+			ret = im_compose(stream, jobid, cookie, IM_ALL_OKAY,
+					event, fromtask, IM_OLD_PROTOCOL_VER);
+			if (ret != DIS_SUCCESS) {
+				arrayfree(argv);
```

Comment:
You can consolidate the "if ret != DIS_SUCCESS" checks to just one block...
ret = im_compose(...);
if (ret == DIS_SUCCESS) diswsi(..);
if (ret == DIS_SUCCESS) diswsi(..);
...
...
if (ret != DIS_SUCCESS) {
     arrayfree(argv);
     arrayfree(envp);
     ....
}